### ISSUE TITLE
ActiveRecord::JDBCError: org.postgresql.util.PSQLException: ERROR: multiple primary keys for table "schema_migrations" are not allowed

### ISSUE THREAD

**Issue by wintersolutions** [jruby/activerecord-jdbc-adapter#734]
When I try my first migration on a new rails 5 app I'm getting the error above.

How to reproduce:

```
rails new foo_bar -database=jdbcpostgresql
cd foo_bar
```

change in Gemfile `gem 'activerecord-jdbcpostgresql-adapter', platforms: :jruby, github: 'jruby/activerecord-jdbc-adapter', branch: 'rails-5'`

```
rails g model foo_bar hello_world
rails db:create
rails db:migrate
=> 
ActiveRecord::StatementInvalid: ActiveRecord::JDBCError: org.postgresql.util.PSQLException: ERROR: multiple primary keys for table "schema_migrations" are not allowed
  Position: 75: CREATE TABLE "schema_migrations" ("version" character varying PRIMARY KEY PRIMARY KEY  PRIMARY KEY)
arjdbc/jdbc/RubyJdbcConnection.java:587:in `execute'
/home/mmonsterkill/.rvm/gems/jruby-9.1.2.0@foo_bar/bundler/gems/activerecord-jdbc-adapter-d9076e990cb2/lib/arjdbc/jdbc/adapter.rb:608:in `_execute'
/home/mmonsterkill/.rvm/gems/jruby-9.1.2.0@foo_bar/bundler/gems/activerecord-jdbc-adapter-d9076e990cb2/lib/arjdbc/jdbc/adapter.rb:584:in `block in execute'
/home/mmonsterkill/.rvm/gems/jruby-9.1.2.0@foo_bar/gems/activerecord-5.0.0/lib/active_record/connection_adapters/abstract_adapter.rb:566:in `block in log'
/home/mmonsterkill/.rvm/gems/jruby-9.1.2.0@foo_bar/gems/activesupport-5.0.0/lib/active_support/notifications/instrumenter.rb:21:in `instrument'
/home/mmonsterkill/.rvm/gems/jruby-9.1.2.0@foo_bar/gems/activerecord-5.0.0/lib/active_record/connection_adapters/abstract_adapter.rb:560:in `log'
/home/mmonsterkill/.rvm/gems/jruby-9.1.2.0@foo_bar/bundler/gems/activerecord-jdbc-adapter-d9076e990cb2/lib/arjdbc/jdbc/adapter.rb:584:in `execute'
/home/mmonsterkill/.rvm/gems/jruby-9.1.2.0@foo_bar/gems/activerecord-5.0.0/lib/active_record/connection_adapters/abstract/schema_statements.rb:283:in `create_table'
/home/mmonsterkill/.rvm/gems/jruby-9.1.2.0@foo_bar/gems/activerecord-5.0.0/lib/active_record/schema_migration.rb:27:in `create_table'
/home/mmonsterkill/.rvm/gems/jruby-9.1.2.0@foo_bar/gems/activerecord-5.0.0/lib/active_record/connection_adapters/abstract/schema_statements.rb:1010:in `initialize_schema_migrations_table'
/home/mmonsterkill/.rvm/gems/jruby-9.1.2.0@foo_bar/gems/activerecord-5.0.0/lib/active_record/migration.rb:1108:in `initialize'
/home/mmonsterkill/.rvm/gems/jruby-9.1.2.0@foo_bar/gems/activerecord-5.0.0/lib/active_record/migration.rb:1005:in `up'
/home/mmonsterkill/.rvm/gems/jruby-9.1.2.0@foo_bar/gems/activerecord-5.0.0/lib/active_record/migration.rb:983:in `migrate'
/home/mmonsterkill/.rvm/gems/jruby-9.1.2.0@foo_bar/gems/activerecord-5.0.0/lib/active_record/tasks/database_tasks.rb:161:in `migrate'
/home/mmonsterkill/.rvm/gems/jruby-9.1.2.0@foo_bar/gems/activerecord-5.0.0/lib/active_record/railties/databases.rake:58:in `block in (root)'
/home/mmonsterkill/.rvm/gems/jruby-9.1.2.0@foo_bar/gems/railties-5.0.0/lib/rails/commands/rake_proxy.rb:13:in `block in run_rake_task'
/home/mmonsterkill/.rvm/gems/jruby-9.1.2.0@foo_bar/gems/railties-5.0.0/lib/rails/commands/rake_proxy.rb:10:in `run_rake_task'
/home/mmonsterkill/.rvm/gems/jruby-9.1.2.0@foo_bar/gems/railties-5.0.0/lib/rails/commands/commands_tasks.rb:51:in `run_command!'
/home/mmonsterkill/.rvm/gems/jruby-9.1.2.0@foo_bar/gems/railties-5.0.0/lib/rails/commands.rb:18:in `<top>'

```
##

**Comment by rebelwarrior** [comment]
I get this with the same gems 
```
$ rails db:migrate
rails aborted!
ActiveRecord::StatementInvalid: ActiveRecord::JDBCError: org.postgresql.util.PSQLException: ERROR: multiple primary keys for table "schema_migrations" are not allowed
  Position: 75: CREATE TABLE "schema_migrations" ("version" character varying PRIMARY KEY PRIMARY KEY  PRIMARY KEY)
arjdbc/jdbc/RubyJdbcConnection.java:587:in `execute'
bin/rails:4:in `<top>'
ActiveRecord::JDBCError: org.postgresql.util.PSQLException: ERROR: multiple primary keys for table "schema_migrations" are not allowed
  Position: 75
arjdbc/jdbc/RubyJdbcConnection.java:587:in `execute'
bin/rails:4:in `<top>'
Tasks: TOP => db:migrate
(See full trace by running task with --trace)
```

**Comment by scytherswings** [comment]
I tried this with a fresh install of jruby-9.0.5.0 on ubuntu 16.04 with these gems installed:

```
$ bundle list
Gems included by the bundle:
  * actioncable (5.0.0)
  * actionmailer (5.0.0)
  * actionpack (5.0.0)
  * actionview (5.0.0)
  * activejob (5.0.0)
  * activemodel (5.0.0)
  * activerecord (5.0.0)
  * activerecord-jdbc-adapter (1.3.20 d9076e9)
  * activerecord-jdbcpostgresql-adapter (1.3.20 d9076e9)
  * activesupport (5.0.0)
  * arel (7.0.0)
  * builder (3.2.2)
  * bundler (1.12.5)
  * concurrent-ruby (1.0.2)
  * erubis (2.7.0)
  * ffi (1.9.14)
  * globalid (0.3.6)
  * i18n (0.7.0)
  * jdbc-postgres (9.4.1206)
  * listen (3.1.5)
  * loofah (2.0.3)
  * mail (2.6.4)
  * method_source (0.8.2)
  * mime-types (3.1)
  * mime-types-data (3.2016.0521)
  * minitest (5.9.0)
  * nio4r (1.2.1)
  * nokogiri (1.6.8)
  * puma (3.4.0)
  * rack (2.0.1)
  * rack-test (0.6.3)
  * rails (5.0.0)
  * rails-dom-testing (2.0.1)
  * rails-html-sanitizer (1.0.3)
  * railties (5.0.0)
  * rake (11.2.2)
  * rb-fsevent (0.9.7)
  * rb-inotify (0.9.7)
  * ruby-maven (3.3.12)
  * ruby-maven-libs (3.3.9)
  * ruby_dep (1.3.1)
  * sprockets (3.6.3)
  * sprockets-rails (3.1.1)
  * thor (0.19.1)
  * thread_safe (0.3.5)
  * tzinfo (1.2.2)
  * tzinfo-data (1.2016.6)
  * websocket-driver (0.6.4)
  * websocket-extensions (0.1.2)
```

And my gemfile:

```
source 'https://rubygems.org'
gem 'rails', '~> 5.0.0'
gem 'ruby-maven'
gem 'activerecord-jdbcpostgresql-adapter'
gem 'activerecord-jdbc-adapter', github: 'jruby/activerecord-jdbc-adapter', branch: 'rails-5'
gem 'puma', '~> 3.0'
gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]

group :development do
  gem 'listen'
end
```

I also get the same result:
```
 $ rails db:migrate
rails aborted!
ActiveRecord::StatementInvalid: ActiveRecord::JDBCError: org.postgresql.util.PSQLException: ERROR: multiple primary keys for table "schema_migrations" are not allowed
  Position: 75: CREATE TABLE "schema_migrations" ("version" character varying PRIMARY KEY PRIMARY KEY  PRIMARY KEY)
arjdbc/jdbc/RubyJdbcConnection.java:587:in `execute'
```

**Comment by kares** [comment]
ActiveRecord 5 is not supported and there's no one working on it, so I encourage anyone to look into it.

**Comment by citizen428** [comment]
I know this issue is somewhat old, but I used the following as a workaround:

1. Switch to use SQL schema format in `config/application.rb`:
    ```
    config.active_record.schema_format :sql
    ```
1. `rake db:structure:dump` to generate ` db/structure.sql`
1. Remove `db/schema.rb`
1. Setup database with `rake db:create` followed by `rake db:structure:load`

Migrations still won't work, but at least you can get a DB up to run specs and see if you have other problems etc.

**Comment by detj** [comment]
There's a `rails-5` branch available. But its not working. Does anyone know how much of an work is it?

**Comment by rosunshrestha** [comment]
Same Problem. I think we have this problem when we use rails 5 with jruby and postgres as database. The possible solution is move rails to version 4.X or author( @kares  ) of acitverecord-jdbc-adapter can help us to use this gem with rails 5.

**Comment by deeppokhrel** [comment]
same problem while migrating my tables to db ,**# ActiveRecord::JDBCError: org.postgresql.util.PSQLException**: ERROR: multiple primary keys for table "schema_migrations" are not allowed. Are any one looking at this issue ??

**Comment by pryanwong** [comment]
Same problem experience while doing a db:setup.
It created all of the other tables but failed at schema_migrations

-- initialize_schema_migrations_table()
rake aborted!

**Comment by skoona** [comment]
This error is very frustration so I took a looked at the code for myself.  Here is what I found so far; a hack but not a true cause.

    `edit gems/activerecord-5.0.1/lib/active_record/connection_adapters/abstract/schema_creation.rb
    - check if already done first
    line 124: to: 'if options[:primary_key] == true && !sql.include?('PRIMARY KEY')'
    line 124: from: 'if options[:primary_key] == true`

    `edit bundler/gems/activerecord-jdbc-adapter-ee68c1881a9b/lib/arjdbc/postgresql/schema_creation.rb
    - comment out lines!
    lines: 17 and 26, from: 'sql << " PRIMARY KEY " '
    lines: 17 and 26, to: '# sql << " PRIMARY KEY " '`

My error had three " PRIMARY KEY " occurrences, I found that one was present when the #visit_ColumnDefinition() is called, it then added one more, then #add_column_options!() added the third.

There is a look of code to follow and I'm not satisfied I've found the true cause.  Hoping these observations will help someone else find a true fix.

My results was, '$rake db:migrate' succeeds as expected.

Now I need to figure out how to package for my personal use, until a fix is implemented.

**Comment by Helsioufy** [comment]
Hey guys, I am getting this error however with a different table:
`ActiveRecord::StatementInvalid: ActiveRecord::JDBCError: org.postgresql.util.PSQLException: ERROR: multiple primary keys for table "ar_internal_metadata" are not allowed`

I specified this in my Gemfile: `gem 'activerecord-jdbcpostgresql-adapter', platforms: :jruby, github: 'jruby/activerecord-jdbc-adapter', branch: 'rails-5'` and still having the same issue.

**Comment by hobodave** [comment]
Fixed this in what I think was a reasonable manner. Conditionally defining methods based on AR versions makes me cringe a little, but the prior art is already there.

https://github.com/jruby/activerecord-jdbc-adapter/pull/781

**Comment by rdubya** [comment]
We are not currently failing any tests related to this, so I am going to mark it as closed with the 50.0 release. Please reopen if it continues to be an issue.
